### DISCUSSION TITLE
How to set UV coordinates per triangle

### DISCUSSION THREAD

**Discussion by royvelich** [pyvista/pyvista#4663] [Q&A] [answered]
Hi,
Let's say that I have a simple 3D cube, so each face is made up of two triangles. How can I set separate UV coordinates per triangle? Is it possible to do it without duplicating vertices?
ChatGPT offered the following solution, which duplicates each vertex of the mesh according to its degree (so essentially the cube becomes a triangle soup). I wonder if there is any other way to achieve it.

```py
import pyvista as pv
import numpy as np

# Define vertices for each triangle (some vertices will be the same spatial coordinates)
vertices = np.array([
    [0, 0, 0],  # triangle 1
    [1, 0, 0],
    [1, 1, 0],

    [0, 0, 0],  # triangle 2
    [1, 1, 0],
    [0, 1, 0], 

    # ... Repeat this pattern for the other 10 triangles
])

# Define the faces (each face is just a list of 3 vertices)
faces = np.array([
    [3, 0, 1, 2],
    [3, 3, 4, 5],
    # ... Repeat this pattern for the other 10 triangles
])

# Create the poly data
cube = pv.PolyData(vertices, faces)

# Define the UV coordinates for each vertex
uv_coords = np.array([
    [0, 0],  # triangle 1
    [1, 0],
    [1, 1],

    [0, 0],  # triangle 2
    [1, 1],
    [0, 1],

    # ... Repeat this pattern for the other 10 triangles
])

# Assign the uv coordinates to the polydata
cube.point_data['Texture Coordinates'] = uv_coords

# Set the active texture coordinates
cube.active_t_coords = 'Texture Coordinates'
```

**Comment by MatthewFlamm** [comment]
Have you looked at the documentation? chatGPT often provides code that looks right, but is incorrect, as in this case.

https://docs.pyvista.org/version/stable/examples/00-load/create-poly.html

**Reply by MatthewFlamm** [reply]
I'm unclear on what behavior you want for UV coordinates.  Is it specifically for textures or interpolation or ?

**Reply by akaszynski** [reply, marked as answer]
@royvelich, what you're describing here is "point-cell" scalars. VTK's data model only supports either point scalars (one set of value(s) per point) or face/cell scalars (one set of value(s) per face). If you want each point on each face to have its own value, you will need to isolate each face.

In this case ChatGPT is correct.
